include/z64cutscene.h

```diff
@@ -3,286 +3,501 @@
+    /* 0x65 */ CS_DEST_DESERT_COLOSSUS_FROM_NABOORU_CAPTURE,
+    /* 0x66 */ CS_DEST_TEMPLE_OF_TIME_FRONT_OF_PEDESTAL, // unused
+    /* 0x67 */ CS_DEST_HYRULE_FIELD_TITLE_SCREEN,
+    /* 0x68 */ CS_DEST_TITLE_SCREEN_DEMO,
+    /* 0x69 */ CS_DEST_GRAVEYARD_SUNS_SONG_PART_2,
+    /* 0x6A */ CS_DEST_ROYAL_FAMILYS_TOMB_SUNS_SONG_PART_3,
+    /* 0x6B */ CS_DEST_GANONS_CASTLE_DISPEL_FOREST_BEAM,
+    /* 0x6C */ CS_DEST_GANONS_CASTLE_DISPEL_WATER_BEAM,
+    /* 0x6D */ CS_DEST_GANONS_CASTLE_DISPEL_SHADOW_BEAM,
+    /* 0x6E */ CS_DEST_GANONS_CASTLE_DISPEL_FIRE_BEAM,
+    /* 0x6F */ CS_DEST_GANONS_CASTLE_DISPEL_LIGHT_BEAM,
+    /* 0x70 */ CS_DEST_GANONS_CASTLE_DISPEL_SPIRIT_BEAM,
+    /* 0x71 */ CS_DEST_GANONS_CASTLE_DISPEL_BARRIER_CONDITONAL,
+    /* 0x72 */ CS_DEST_HYRULE_FIELD,
+    /* 0x73 */ CS_DEST_HYRULE_FIELD_FROM_IMPA_ESCORT,
+    /* 0x74 */ CS_DEST_FROM_RARU_CONDITIONAL,
+    /* 0x75 */ CS_DEST_HYRULE_FIELD_CREDITS_SKY,
+    /* 0x76 */ CS_DEST_GANON_BATTLE_TOWER_COLLAPSE,
+    /* 0x77 */ CS_DEST_ZELDAS_COURTYARD_RECEIVE_LETTER
+} CutsceneDestination;
 
 typedef struct {
     /* 0x00 */ u16 base;
     /* 0x02 */ u16 startFrame;
     /* 0x04 */ u16 endFrame;
-} CsCmdBase; // size = 0x6
+    /* 0x06 */ u8 unused_06[0x2];
+} CsCmdGeneric; // size = 0x8
 
 typedef struct {
-    /* 0x00 */ u8  unk_00;
-    /* 0x01 */ u8  setting;
+    /* 0x00 */ u16 type;
     /* 0x02 */ u16 startFrame;
     /* 0x04 */ u16 endFrame;
-} CsCmdEnvLighting; // size = 0x6
+    /* 0x06 */ u8 unused_06[0x2A];
+} CsCmdMisc; // size = 0x30
 
 typedef struct {
-    /* 0x00 */ u8  unk_00;
-    /* 0x01 */ u8  sequence;
-    /* 0x02 */ u16 startFrame;
-    /* 0x04 */ u16 endFrame;
-} CsCmdMusicChange; // size = 0x6
+    /* 0x00 */ u8 unused_00[0x1];
+    /* 0x01 */ u8 settingPlusOne;
+    /* 0x02 */ u16 frame;
+    /* 0x04 */ u8 unused_04[0x2C];
+} CsCmdLightSetting; // size = 0x30
+
+typedef struct {
+    /* 0x00 */ u8 unused_00[0x1];
+    /* 0x01 */ u8 seqIdPlusOne;
+    /* 0x02 */ u16 frame;
+    /* 0x04 */ u8 unused_04[0x2C];
+} CsCmdStartSeq; // size = 0x30
+
+typedef struct {
+    /* 0x00 */ u8 unused_00[0x1];
+    /* 0x01 */ u8 seqIdPlusOne;
+    /* 0x02 */ u16 frame;
+    /* 0x04 */ u8 unused_04[0x2C];
+} CsCmdStopSeq; // size = 0x30
 
 typedef struct {
     /* 0x00 */ u16 type;
     /* 0x02 */ u16 startFrame;
     /* 0x04 */ u16 endFrame;
-} CsCmdMusicFade; // size = 0x6
+    /* 0x06 */ u8 unused_06[0x2A];
+} CsCmdFadeSeq; // size = 0x30
 
 typedef struct {
-    /* 0x00 */ u16 unk_00;
+    /* 0x00 */ u8 unused_00[0x2];
     /* 0x02 */ u16 startFrame;
     /* 0x04 */ u16 endFrame;
     /* 0x06 */ u8  unk_06;
     /* 0x07 */ u8  unk_07;
     /* 0x08 */ u8  unk_08;
-} CsCmdUnknown9; // size = 0xA
+    /* 0x0A */ u8 unused_0A[0x2];
+} CsCmdRumble; // size = 0xC
 
 typedef struct {
-    /* 0x00 */ u16 unk_00;
-    /* 0x02 */ u16 startFrame;
-    /* 0x04 */ u16 endFrame;
+    /* 0x00 */ u8 unused_00[0x2];
+    /* 0x02 */ u16 frame;
+    /* 0x04 */ u8 unused_04[0x2];
```

Comment:
```suggestion
    /* 0x00 */ u8 unused_00[0x2];
    /* 0x02 */ u16 startFrame;
    /* 0x04 */ u16 endFrame;
```

Yea I've gotta put my foot down on this.

I don't believe it is in the spirit of the project to deliberately obfuscate a data type's underlying structure simply because the engine itself never uses those properties. Not at the struct level. I know you probably know this already, but when you look through the data for various cutscenes throughout the game, it seems extremely likely that + 0x04 is well first of all meant to be either a s16 or an u16, and second by comparing it to the other commands it's meant to be the end frame for all commands. Maybe you know, but I don't know of a single instance where +0x04 seems like an unreasonable or illogical end frame for a command.

Reply:
It sounds like youd rather us go down the route of using general command structs instead of a unique struct for each one. which is a fair opinion to have. Though there are some issues with this (see comment below about data types)

However I do not understand at all your point about this being worse for modding. With this approach of the structs being seperate for each command, if a field isnt relevant at all, then it being named doesnt really matter. If anything its more confusing as a user to see an end frame argument and have it not do anything.
Its not muddying the interface, its just making it clear what matters. If someone wants to add an end frame to a command that doesnt currently use it, they can add the field in.

Reply:
I do question where limit should be when it comes to "retrofitting" new designs on top of the old ones (I believe the DOOR_ACTOR_BASE macro is example of a retrofit solution), because even when decomp is "done" it's never going to be the ideal starting point for a new mod. There are so many systems that can be rewritten to be faster or more intuitively or be stripped out entirely because they're not doing what they need to do, and then there's also the desire for cross compatibility with MM as well. It's not that retrofitting shouldn't be done... rather it's that the retrofit should still roughly express the underlying design behind it, without being overly complex given that complex designs will be scrapped by modders anyway.

It's not so much that I'm not fully confident that specialized structs are better than generic structs (I'm leaning 90% towards specialized structs), it's that I can't stand behind an implementation that deliberately chooses not to correctly implement all known properties (including unused ones) to the best of our ability just because doing so might inconvenience someone new to the codebase. I don't feel this is the appropriate thing to be doing in decomp, nor do I feel that these structs are well designed for doing so.

As for startFrame/endFrame specifically, my point is that after looking at the data I believe there is suppose to be a common access point (i.e. interface) for retrieving the startFrame and endFrame for all cutscene commands, something your pr will ruin by not maintaining an endFrame for all structs. From my personal experience with the creation of the shortened cutscenes in Beta Quest, this is an incredibly beneficial interface to have, since for example it makes writing a "trim" function that little bit easier to code.

Reply:
I just dont follow how having an endFrame named as such when it doesnt behave that way is beneficial. It only adds room for confusion. 
If someone wanted to add in an endFrame field to a command that didnt have one, it is very easy to add it in yourself. And I dont think there is anything to gain from the values that they have as "endFrame" for those commands. Most of them are just startFrame + 1 cause these commands do their job in 1 frame anyway

Reply:
because it's an INTERFACE, likely meant for tooling external to the game itself. You aren't obligated to use all of the properties in an interface, it just needs to exist for interoperability. The whole point here is to define a common structure for defining new commands; it's not simply a matter of "oh we can add an endFrame later" when someone else creates a new command that doesn't have an endFrame there. This is why we have structs in the first place, to reserve memory for specific purposes, since otherwise they would get clobbered. 

But let's go even deeper into the design of these structs. Explain to me why it's better design to access the property of e.g. CsCmdLightSetting at +0x04 using a byte array.

Reply:
im not saying to access it via the byte array, im saying to add in a variable of your choosing at that point, whenever you want to use it. it can be an endFrame if you want to add that functionality, or anything. Its free space as far as im concerned, cause it is not used at all in the current implementation of the command

if you want them to be stuck to the same common structure then you should go all the way and use common structs. but as ive articulated already, I think its worse overall and doesnt add any benefit

Reply:
why should I be the one responsible for creating types that should already exist? you're the modder, you can remove what you think is useless or not.

Reply:
because it *is* useless
it means nothing in the context of this command unless its actually added in. if you want to add functionality for it, you can also be expected to add the variable needed to hold that information

Reply:
we'll prob just need other opinions cause I dont think we'll agree on this topic.
But I do still feel like you cant have it both ways. The whole points of spcialized structs here is to make them reflect their proper usage. If you want it to state that the useless variable is an end frame, you have to commit to full on common structs

Reply:
This is the general direction for the changes I want to see in z_demo, though not all necessarily in this PR. It's not merge ready because I gave up when I saw csdis was a thing, but it compiles and O.K.s See z64cutscene_commands.h, z64cutscene.h

https://github.com/fig02/oot/compare/z_demo...mzxrules:oot:z_demo

All CsCmds start with `u16 action`, `u16 startFrame`, `u16 endFrame` and so I've at least made them follow this structure. I opted not to make them directly use a common struct because I wasn't feeling it. I've also converted the u8 "action" properties to u16, since it eliminates all instances of "unused" variables in the struct definition at the cost of 3 casts. Technically this also fixes the UB with a few of the macros (CS_LIGHT_SETTING), but I imagine that was going to be fixed with future ZAPD changes.

CsCmdMisc, CsCmdLightSetting, CsCmdStartSeq, CsCmdStopSeq, CsCmdFadeOutSeq are clearly all `CsCmdActorCue`s. I hastily added in a new `CsCmdCommonCue full` to tie the struct relationship in code so that the internal properties can still be read from, but I'm not strongly committed to this implementation. 

With the above being `CsCmdActorCue`s, their macros can be simplified to e.g. `CS_MISC(action, startFrame, endFrame, posX, posY, posZ)` (ignore name changes and missing -1 operations). Position seems like the most likely property being stored here, and I think it's plausible for cutscenes to have been edited in an object based editor.

Reply:
Would you remove `CS_MISC`, `CS_LIGHTING`, ... to replace them with `CS_ACTOR_CUE` ? (or `CS_ACTOR_CUE` renamed to something else)

Reply:
The `CS_MISC`, `CS_LIGHTING`, etc. macros would still exist, since they simplify the initialization of the full `CS_ACTOR_CUE` (CS_MISC is 6 args vs `CS_ACTOR_CUE`'s 15). `CS_ACTOR_CUE` would need to be renamed, but the best I can think of would be something like `CS_CMD_COMMON_CUE` to match the `CsCmdCommonCue` struct I threw in at the last minute.